waku/v2/node/peer_manager/waku_peer_store.nim

```diff
@@ -62,12 +67,41 @@ type
     connectedness*: Connectedness
     disconnectTime*: int64
     origin*: PeerOrigin
-    direction*: Direction
+    direction*: PeerDirection
+    lastFailedConn*: Moment
+    numberFailedConn*: int
 
 ##################
 # Peer Store API #
 ##################
 
+proc canBeConnected*(peerStore: PeerStore,
+                     peerId: PeerId,
+                     initialBackoffInSec: int,
+                     backoffFactor: int): bool =
+  # Returns if we can try to connect to this peer, based on past failed attempts
+  # It uses an exponential backoff. Each connection attempt makes us
+  # wait more before trying again.
+  let failedAttempts = peerStore[NumberFailedConnBook][peerId]
+
+  # if it never errored, we can try to connect
+  if failedAttempts == 0:
+    return true
+
+  # If it errored we wait an exponential backoff from last connection
+  # the more failed attemps, the greater the backoff since last attempt
+  let now = Moment.init(getTime().toUnix, Second)
+  let lastFailed = peerStore[LastFailedConnBook][peerId]
```

Comment:
```suggestion
  let lastFailed = peerStore.getLastFailedPeer(peerId)
```
or similar, which would allow us to change the underlying datastructure if required in the futute

Reply:
I agree with this, but I'm just trying to follow nimlibp2p peerstore https://github.com/status-im/nim-libp2p/blob/unstable/libp2p/peerstore.nim#L148  pattern, of using custom books.

Doing this will require to have 6-7 new getter functions, with just one line of code, and not sure I see the benefit right now.

But I'm totally open for suggestions.

Reply:
Sounds good! just something to keep in mind if we decide to have more complex functionality later.